scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/KeyGroupIterator.java

```diff
@@ -17,69 +17,92 @@
 
 package org.apache.beam.sdk.extensions.smb;
 
+import java.util.Collections;
 import java.util.Comparator;
 import java.util.Iterator;
+import java.util.List;
 import java.util.NoSuchElementException;
 import java.util.function.Function;
+import java.util.stream.Collectors;
+
 import org.apache.beam.sdk.values.KV;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.PeekingIterator;
 
 // FIXME: current limitation: must exhaust Iterator<ValueT> before starting the next key group
 class KeyGroupIterator<KeyT, ValueT> implements Iterator<KV<KeyT, Iterator<ValueT>>> {
-  private final PeekingIterator<ValueT> iterator;
+  private final List<PeekingIterator<ValueT>> iterators;
   private final Function<ValueT, KeyT> keyFn;
   private final Comparator<KeyT> keyComparator;
 
   private Iterator<ValueT> currentGroup = null;
 
+  // FIXME: remove
   KeyGroupIterator(
       Iterator<ValueT> iterator, Function<ValueT, KeyT> keyFn, Comparator<KeyT> keyComparator) {
-    this.iterator = Iterators.peekingIterator(iterator);
+    this.iterators = Collections.singletonList(Iterators.peekingIterator(iterator));
+    this.keyFn = keyFn;
+    this.keyComparator = keyComparator;
+  }
+
+  KeyGroupIterator(
+      List<Iterator<ValueT>> iterators,
+      Function<ValueT, KeyT> keyFn,
+      Comparator<KeyT> keyComparator) {
+    this.iterators =
+        iterators.stream().map(Iterators::peekingIterator).collect(Collectors.toList());
     this.keyFn = keyFn;
     this.keyComparator = keyComparator;
   }
 
   @Override
   public boolean hasNext() {
     checkState();
-    return iterator.hasNext();
+    return iterators.stream().anyMatch(it -> it.hasNext() && it.peek() != null);
```

Comment:
I'm not sure this 100% correct. The `!= null` condition is basically treating `null` key as not a real key but it also seems that our E2E test doesn't cover cases with multiple `null` keys that might mess up the iterator state?

Reply:
This might be a red herring and caused by the bug in `JsonFileOperations`.